[BEAM-7304] Twister2 Beam runner

pulasthi opened the pull request:
This pull includes a new runner for Beam. The runner is based on Twister2 [1]. The current implementation of the runner supports batch pipelines. 


[1] https://twister2.org/

R: @kennknowles @iemejia 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

iemejia commented:
retest this please

iemejia commented:
retest this please

pulasthi commented:
> retest this please

was that comment meant to trigger the tests?

pulasthi commented:
retest this please

pulasthi commented:
retest this please

iemejia commented:
Yes exact. But you cannot trigger the tests at the moment because of some issues we have You can ask however in this PR or better in the mailing list so someone else does.

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
arrghh obbey Jenkins or I will put you back in your cage !!!

iemejia commented:
Mmm somethings odd, sometimes we need to retrigger the PRs more than one time, but now looks like more broken than usual.

pulasthi commented:
It looks like its retesting ignoring the last commit which fixes some of the earlier check failures

pulasthi commented:
@iemejia the failure details are showing errors from a .gcp package which are not related to this pull, I did not notice any other failures. Is this due to some other commit or am I missing something?

iemejia commented:
Yes that sadly is common, for the moment ignore such errors if they repeat it means they are 'flaky tests', please report in the mailing list so someone takes a look.

I took an ultra quick look at your PR. It is currently not running the tests because you forgot to add the twister2 runner in the settings.gradle file.

Please also try to fix the name of the classes `Twister2PiplineExecutionEnvironment` and `Twister2PiplineResult` so they are `Pipeline` instead of `Pipline`.

I also look at the Validates runner tests that are excluded and I would strongly recommend you to try to make the UsesParDoLifecycle tests pass, those matter. If your ParDo translation is correct (and you respect the lifecycle) you will easily be able to remove the `UsesBoundedSplittableParDo` exclusion too because they are expanded into normal ParDos by default.

Finally worth also to try to get the `UsesSchema` tests passing because that should not be difficult, that's just passing around some object you are probably not doing in the translation and the Twister2 users will win the guarantee that SQL/Schema-based PCollections are working.

iemejia commented:
Forgot to mention, can you please do a rebase of your branch, your commits look a bit messed up.

iemejia commented:
WindowingTests are failing in the ValidatesRunner suite. So take a look at those. I would be a good idea to add to this PR a Jenkins invocation of the twister2 validatesrunner tests see for reference https://github.com/apache/beam/blob/master/.test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Flink.groovy

pulasthi commented:
@iemejia Thanks for your feedback. Fixed the class name typos, rebased the branch and added the post-commit script as requested. I will work on the test cases that you have pointed out next.

Regarding the settings.gradle file, I did add the following entry "include ":runners:twister2" into that file, is there another entry that I need to make so tests are executed?

iemejia commented:
For the settings.gradle that's exactly what you had to do, now it will be part of the precommit run, let's see if it goes well.

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
For the ValidatesRunner job, we do have to execute a seed job than 'installs' it and starting with that it can be triggered by the "Run Twister2 ValidatesRunner" phrase. Let me see if that works too.

iemejia commented:
retest this please

iemejia commented:
Run Seed Job

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Nice for the fixes! (Hope with those you can get BoundedSDF in too) Apparently the VR script is working too (not necessarily the tests passing but that I know you are still checking). Can you also please add the runner into `.github/PULL_REQUEST_TEMPLATE.md` with the respective Twister2 status link, notice that they are in lexicographical order, so put the column at the end.

pulasthi commented:
@iemejia that's nice to hear. I am getting an "UnsupportedOperationException" from a call to the "restrictionTracker" method in the SimpleDoFnRunner when I try to run the BoundedSDF tests. I get this in multiple tests, WIll keep working on that, will ask the dev list if I get stuck with that issue.

I added to the pull request template as you mentioned. I assume i would have to submit an ICLA since this is a large contribution.

alexvanboxel commented:
You can add the `twister2` to .gitlab/autolabeler.yaml . I've add it to the gitlab labels. Please add it to the runner section (and keep alphabetic order into account).

iemejia commented:
@pulasthi Yes the ICLA is a really good idea given the size of the contribution.

pulasthi commented:
@iemejia Just added a fix so the 'UsesBoundedSplittableParDo' tests are passing. Regarding the ICLA I assume I need to mention Apache beam under "(optional) notify project" ?

pulasthi commented:
@iemejia @kennknowles would you be able to review the changes, and maybe trigger the tests and build again.

iemejia commented:
retest this please

iemejia commented:
Run Java PreCommitRun Java PreCommit

iemejia commented:
Run Java PreCommit

iemejia commented:
the failing JavaPortabilityApiJava11 PreCommit job is an error and will dissapear if you rebase this PR.

pulasthi commented:
@iemejia Thanks, I just rebased the PR, please let me know if anything else is needed from my end to proceed with the review.

iemejia commented:
retest this please

iemejia commented:
retest this please

pulasthi commented:
@iemejia the java pre-commit still seems to be failing but because of other reasons.

iemejia commented:
If you don't mind making a single commit probably will make things more manageable (for the rebase), but we can also do this at the end, do as you prefer.

iemejia commented:
@kennknowles do you have some time to do a review round on this one? if not can you suggest another additional reviewer? It looks pretty solid already I am optimistic that it could make it for 2.21.0

pulasthi commented:
@iemejia rebased the pull to a single commit. Indeed a single commit would be easier to manage.

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
Jenkins don't be like that one day you will have to deal with people like you please!

iemejia commented:
retest this please

pulasthi commented:
@iemejia sorry I think my last push reset the test run

iemejia commented:
retest this please

iemejia commented:
@pulasthi can you please do a `git pull origin master --rebase` and push force on this one, it seems that when you did the PR something was failing and the full tests are not running. Sorry for so many inconviniences.

iemejia commented:
Oh ignore my previous comment it worked now. ARRRGHHH Jenkins!

pulasthi commented:
@iemejia Thanks, no inconvenience at all. Hopefully, we can get this into 2.21 as you mentioned

pulasthi commented:
@iemejia just wanted to check if you were able to review the code. And please let me know if you need anything from my end

iemejia commented:
Not finished yet, sorry I had to deal first with the unintended consequences of the whole coronavirus lockdown. I will get to it ASAP. Sorry again.

pulasthi commented:
No worries, I assumed that might have been the case.

pulasthi commented:
@iemejia just wanted to check if you got some free time to look into this, hope you are staying safe

pulasthi commented:
@iemejia Hope you are doing well. I just wanted to follow up with you if you had time to work on the pull request.

iemejia commented:
Hi @pulasthi my BIGGEST excuses for my delay. I have not forgotten about this one it is just in bag of guilt. Expect some news beginning of the week. In the meantime thanks for rebasing and updating it.

iemejia commented:
retest this please

pulasthi commented:
@iemejia Thanks for the update. I just fixed a couple of errors that were caused by changes in the beam master.

pulasthi commented:
@iemejia Would you be able to trigger the checks again

iemejia commented:
retest this please

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

iemejia commented:
Run Seed Job

pulasthi commented:
@iemejia I checked the errors in the java build and I am not sure where it originates from, it does seem to be from the changes in this pull request

iemejia commented:
Yes they look unrelated

iemejia commented:
Run Java PreCommit

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Run Java PreCommit

iemejia commented:
Run Python PreCommit

iemejia commented:
Run Python PreCommit

iemejia commented:
Run Python PreCommit

iemejia commented:
Run Twister2 ValidatesRunner

TheNeuralBit commented:
Run Python PreCommit

TheNeuralBit commented:
Run Twister2 ValidatesRunner

TheNeuralBit commented:
https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2_PR/ says "This project is currently disabled"

TheNeuralBit commented:
Run Seed Job

TheNeuralBit commented:
Run Twister2 ValidatesRunner

TheNeuralBit commented:
huzzah! when I logged in to jenkins an "enable project" button appeared

RyanSkraba commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2StreamTranslationContext.java:
```diff
@@ -0,0 +1,21 @@
+package org.apache.beam.runners.twister2;
+
+/** Twister2StreamTranslationContext. */
+public class Twister2StreamTranslationContext {}
```
Is this file a mistake?  There's already `Twister2StreamingTranslationContext` so it could lead to confusion with the similar name.

pulasthi replied:
Yes it is :smile: , there was a duplicate class fixed it.

RyanSkraba commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+
+/** Internal implementation of the Beam runner for Apache Flink. */
```
Looks like Twister2 to me!  :smile: (Fix comment in all package-info.java)

pulasthi replied:
wonder how i missed that, i was sure i searched for spark and flink :smile: . Fixed all the instances

RyanSkraba commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/translators/functions/AssignWindowsFunction.java:
```diff
@@ -0,0 +1,119 @@
+    implements ComputeCollectorFunc<WindowedValue<T>, Iterator<WindowedValue<T>>> {
+  private static final Logger LOG = Logger.getLogger(AssignWindowsFunction.class.getName());
+
+  private transient boolean isInitilized = false;
```
Typo `isInitialized` (throughout the code).

pulasthi replied:
Thank you for catching that, fixed all the occurrences

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2LegacyRunner.java:
```diff
@@ -0,0 +1,339 @@
+   *
+   * @param options
+   */
+  private void zipFilesToStage(Twister2PipelineOptions options) {
```
@RyanSkraba Thank you for reviewing the code, It is good to get another set of eyes to notice mistakes that one makes. 

A single zip file is being created for each test since Twister2 treats each test as a new job, the zip file contains the dependency files that are needed by the run. I will look into a way to delete the zip files created by tests that have completed so it does not fill up the disk space.

pulasthi replied:
I found a solution with forkEvery, I set the temp files created to delete on exit and since the JVM changes for every test class the old temp files will get deleted automatically.

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2PipelineOptions.java:
```diff
@@ -0,0 +1,93 @@
+  @Description("set unique application name for Twister2 runner")
+  void setApplicationName(String name);
+
+  String getApplicationName();
```
I initially added them for completeness, but ended up not using some of them. removed the unused options since they can be added if needed later.

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2PipelineOptions.java:
```diff
@@ -0,0 +1,93 @@
+
+/** Twister2PipelineOptions. */
+public interface Twister2PipelineOptions extends PipelineOptions, StreamingOptions {
+  @Description("set unique application name for Twister2 runner")
```
Was not aware of that convention, thanks for pointing that out. reformatted according this rule

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2LegacyRunner.java:
```diff
@@ -0,0 +1,339 @@
+    Twister2PipelineResult result = new Twister2PipelineResult();
+    if (jobState.getJobstate() == DriverJobState.FAILED
+        || jobState.getJobstate() == DriverJobState.RUNNING) {
+      throw new RuntimeException("Pipeline execution failed", jobState.getCause());
```
This is something that needs to be fixed form the Twister2 side first, currently the submit method returns once the job is complete, and if it returns with the state as "RUNNING" it means that the job terminated and did not complete. This can be fixed once the issue is fixed on the Twister2 side. I hope keeping it as is for now is OK for now.

RyanSkraba replied:
Thanks for the clarification!

pulasthi commented on runners/twister2/build.gradle:
```diff
@@ -0,0 +1,89 @@
+}
+description = "Apache Beam :: Runners :: Twister2"
+
+repositories {
```
Thanks for catching that, removed it

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2LegacyRunner.java:
```diff
@@ -0,0 +1,339 @@
+  /** Provided options. */
+  private final Twister2PipelineOptions options;
+
+  public Twister2LegacyRunner(Twister2PipelineOptions options) {
```
Changed to protected

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2PipelineResult.java:
```diff
@@ -0,0 +1,58 @@
+
+  @Override
+  public State waitUntilFinish(Duration duration) {
+    return State.DONE;
```
Those operations are not supported currently, I missed to mark them as such. Fixed them

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/BeamBatchWorker.java:
```diff
@@ -0,0 +1,162 @@
+  private static final String LEAVES = "leaves";
+  private static final String GRAPH = "graph";
+  private HashMap<String, BatchTSet<?>> sideInputDataSets;
+  private Set<TSet> leaves;
```
I might have to take a little time on this and make sure i get the type information correct. Would it be OK to create improvement issue for this for now, so i can get back to it later after the code is merged?

RyanSkraba replied:
Of course!  Objectively, this is **_only_** a preference, I'm OK for leaving it for a subsequent improvement/cleanup.

RyanSkraba commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+
+/** Internal implementation of the Beam runner for Apache Twister2. */
```
I am SO sorry, such a minor nitpick but probably the right time to fix it!  You probably want to remove or move Apache here.
```suggestion
/** Internal implementation of the Beam runner for Twister2. */
```

pulasthi replied:
Fixed it, it is good that you caught that, that would go unfixed for a long time otherwise

RyanSkraba commented:
I ran the `validatesRunner` tests again and the partition /tmp directory wasn't filled with zip files!  Thanks for your speedy attention, LGTM!

pulasthi commented:
@RyanSkraba Thank you so much for your review and feedback

pulasthi commented:
@iemejia Would we be able to merge this into the master so it would be included in the 2.23.0 release? let me know what else needs to happen on my end to proceed.

iemejia commented:
@pulasthi technically there are not many major things. (I may comment on those but don't worry for those much). Currently the only concern I have is about the [bus factor](https://en.wikipedia.org/wiki/Bus_factor) of this contribution. Is there anyone apart of you able to contribute fixes/support for this runner in the future?

Also given the size of the contribution and the fact that it comes from a University (aka corporation) you need to send a signed CCLA https://www.apache.org/licenses/cla-corporate.pdf 
More info on why here https://www.apache.org/licenses/contributor-agreements.html

Once you have the signed document from your University please send a copy to secretary@apache.org and private@beam.apache.org

If you have still questions about this legal part please ask those to private@beam.apache.org

Thanks

pulasthi commented:
@iemejia Looking forward for your feedback. And about the maintainability question, the Twister2 project has 10-15 active contributors at the moment who can take over the responsibility to maintain the contribution if I am not able to work on it somehow. Several of them know the codebase (of the Twister2 Runner) in detail, so it should not be an issue.

We are also planning to join the Apache incubator in the near future after a couple of more features that we think are important are completed. The Twister2 Github page seems a little inactive these days because we are working on the Twisterx ( https://github.com/DSC-SPIDAL/twisterx ) project which will be merged into the Twister2 codebase in the coming weeks :). The beam runner is a major aspect that we plan to update and develop in the future.

I hope that answers your concerns to some level. I will also work with the University and get the required CCLA as requested.

Best Regards,
Pulasthi

pulasthi commented:
@iemejia I got the CCLA document from the university and sent to secretary@apache.org and private@beam.apache.org, I did not get an ack back yet, Do i need to attach the document here as well?

iemejia commented:
We received it. We just have to wait for the ack from the people at the ASF. We will let you know. Thanks!

iemejia commented:
There is a list of follow up tasks to be done mostly documentation related but let's do those in a different PR to not delay this one anymore.

Can you please add the missing website documentation, please create and add the runner in the following files.
It is really important that you add the runner to the capability matrix with the data corresponding to the passing ValidatesRunner tests aka the runner capabilities.
```
website/www/site/content/en/documentation/_index.md
website/www/site/content/en/documentation/runners/twister.md
website/www/site/content/en/get-started/beam-overview.md
website/www/site/content/en/roadmap/twister-runner.md 
website/www/site/data/capability_matrix.yaml
website/www/site/layouts/partials/section-menu/en/runners.html
website/www/site/layouts/partials/section-menu/en/roadmap.html
```

iemejia commented:
We need to add the quickstarts and the runner maven archetypes (this can be also another separate PR please).

This basically help us validate that the runner can run WordCount and we don't have regressions in new releases. See the files for more info:
```
sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml 
release/build.gradle
release/src/main/groovy/quickstart-java-twister.groovy
release/src/main/scripts/run_rc_validation.sh
website/www/site/content/en/get-started/quickstart-java.md
website/www/site/content/en/contribute/release-guide.md 
```

For extra awareness maybe you can take a look at the files touched by the recent PRs that removed runners just to be aware of what pieces can be still missing (#11960 and #11990) of course try to copy the way the existing runners do it (instead of the removed ones)

pulasthi commented:
@iemejia Sorry about the delay to respond was stuck on some other work. I will get these done ASAP and update you, Thank you for your time

pulasthi commented:
@iemejia Just to clarify, the documentation work mentioned in the last two comments are to be addressed in separate PR's right? for this PR i need to resolve the code related comments you have added, add to Changes.md and rebase the PR, right?

Best
Pulasthgi

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/Twister2LegacyRunner.java:
```diff
@@ -0,0 +1,342 @@
+ * to a Twister2 Plan and then executing them either locally or on a Twister2 cluster, depending on
+ * the configuration.
+ */
+public class Twister2LegacyRunner extends PipelineRunner<PipelineResult> {
```
Yes, that was the intention of using the legacy part, I can rename it if needed? what do you think would be the best approach?

iemejia replied:
I think `Legacy` has subtle connotations e.g. being old and 'battle tested' but also difficult to upgrade which in this case do not apply, maybe we can call it classic just to be consistent with the terminology we use now to refer to the non-portable Spark/Flink runners e.g. `Twister2ClassicRunner` or just let rename it `Twister2Runner` and call a future portable based runner `Twister2PortableRunner`. So up to you to decide which of the two you prefer.

pulasthi replied:
Renamed it to Twister2Runner

pulasthi commented on runners/twister2/build.gradle:
```diff
@@ -0,0 +1,89 @@
+        exclude group: 'com.esotericsoftware.kryo', module: 'kryo'
+    }
+    compile project(":runners:java-fn-execution")
+    compile "org.twister2:api-java:0.6.0"
```
Done, thank you for pointing that out

pulasthi commented on runners/twister2/src/main/java/org/apache/beam/runners/twister2/translators/functions/ByteToWindowFunctionPrimitive.java:
```diff
@@ -0,0 +1,115 @@
+    }
+    WindowedValue<KV<K, V>> element;
+    if (value != null) {
+      element =
```
Corrected that

pulasthi commented:
@iemejia Added the changes you mentioned, I am not entirely sure if the entry I made in the README.md is what you meant. Let me know if any changes need to be done.

iemejia commented:
> @iemejia Just to clarify, the documentation work mentioned in the last two comments are to be addressed in separate PR's right? for this PR i need to resolve the code related comments you have added, add to Changes.md and rebase the PR, right?
> 
> Best
> Pulasthgi

Yes exactly I would like if you have the time/availability to get those merged 'rapidly' after this one.

iemejia commented on CHANGES.md:
```diff
@@ -51,6 +51,7 @@
 
 ## Highlights
 
+Twister2 Runner ([BEAM-7304](https://issues.apache.org/jira/browse/BEAM-7304))
```
nit: Use the list symbol:

`* New` Twister2 Runner ....

iemejia commented:
Ping me when the rename is done and I will run the tests and then we will be ready to merge.

iemejia commented:
Just a reminder that branch cut is next wednesday (not pressure just a reminder just in case)

pulasthi commented:
@iemejia Thanks for the heads up, I will complete the changes today and ping you back.

pulasthi commented:
@iemejia Just made the changes, Sorry about the delay, Let me know if anything else is needed from my end, I will start work on the document PR's

iemejia commented:
Run Seed Job

iemejia commented:
Run Python PreCommit

iemejia commented:
Run PythonFormatter PreCommit

iemejia commented:
Run Twister2 ValidatesRunner

iemejia commented:
Python errors are unrelated to this PR so merging

pulasthi commented:
@iemejia Thank you so much for the time you put in. I will create a JIRA for that and start working on the documentation you mentioned and create a JIRA to track the progress on that work too. I hope to keep contributing and soon get the Portable runner and streaming runner too.
